### PR TITLE
Fix -b

### LINKED ISSUE #220: -b is broken in plot_krun_results

### PR DESCRIPTION
**Pull request by vext01**
Trying to fix -b. Fixes #220 

I currently have a crash:

```
python2.7 -mpdb /home/vext01/research/warmup_experiment/bin/plot_krun_results b3/warmup_results.json.bz2  -b bencher3:fannkuch_redux:HHVM:default-php:0 -b bencher3:fasta:Hotspot:default-java:0 -o /tmp/xxx.pdf
[2] > /home/vext01/research/warmup_experiment/bin/plot_krun_results(4)<module>()
-> """
(Pdb++) c
Saving results to: /tmp/xxx.pdf
Using matplotlib version 1.4.2 with backend agg
Charting with sliding window size: 200, xlimits: None
Loading: b3/warmup_results.json.bz2
No VM instrumentation data is available.
Adding run sequence to  fasta:Hotspot:default-java bencher3
Adding run sequence to  fannkuch_redux:HHVM:default-php bencher3
Plotting page 01 of 02. On this page, 2 plots will be arranged in 1 rows and 2 columns.
Saved: /tmp/xxx.pdf
Traceback (most recent call last):
  File "/usr/lib/python2.7/pdb.py", line 1314, in main
    pdb._runscript(mainpyfile)
  File "/usr/lib/python2.7/pdb.py", line 1233, in _runscript
    self.run(statement)
  File "/usr/lib/python2.7/bdb.py", line 400, in run
    exec cmd in globals, locals
  File "<string>", line 1, in <module>
  File "/home/vext01/research/warmup_experiment/bin/plot_krun_results", line 4, in <module>
    """
  File "/home/vext01/research/warmup_experiment/bin/plot_krun_results", line 252, in main
    inset_xlimit)
  File "/home/vext01/research/warmup_experiment/bin/plot_krun_results", line 602, in draw_page
    titles[index], x_bounds, [y_min, y_max], window_size,
IndexError: list index out of range
Uncaught exception. Entering post mortem debugging
Running 'cont' or 'step' will restart the program
[6] > /home/vext01/research/warmup_experiment/bin/plot_krun_results(602)draw_page()
-> titles[index], x_bounds, [y_min, y_max], window_size,
[6] > /home/vext01/research/warmup_experiment/bin/plot_krun_results(602)draw_page()
-> titles[index], x_bounds, [y_min, y_max], window_size,
(Pdb++) index
1
(Pdb++) titles
['Fannkuch Redux, HHVM, Linux1/i7-4790K, Process execution #1']
(Pdb++) u
[5] > /home/vext01/research/warmup_experiment/bin/plot_krun_results(252)main()
-> inset_xlimit)
(Pdb++) all_subplot_titles
[['Fannkuch Redux, HHVM, Linux1/i7-4790K, Process execution #1'], ['Fasta, Hotspot, Linux1/i7-4790K, Process execution #1']]
```

My head hurts. Help :cactus:

### CODE REVIEW

**Comment by snim2** [comment]
I think `--wallclock-only` with `-b` may also be broken on this branch:

``` python
../../bin/plot_krun_results -m -t --wallclock-only -b bencher5:binarytrees:C:default-c:0 -o minusb.pdf warmup_results.json.bz2 
/home/snim2/.local/lib/python2.7/site-packages/matplotlib/__init__.py:872: UserWarning: axes.color_cycle is deprecated and replaced with axes.prop_cycle; please use the latter.
  warnings.warn(self.msg_depr % (key, alt_key))
Saving results to: minusb.pdf
Using matplotlib version 1.5.1 with backend agg
Charting with sliding window size: 200, xlimits: None
Loading: warmup_results.json.bz2
No VM instrumentation data is available.
Adding run sequence to  binarytrees:C:default-c bencher5
Traceback (most recent call last):
  File "../../bin/plot_krun_results", line 1087, in <module>
    options.changepoints)
  File "../../bin/plot_krun_results", line 878, in get_data_dictionaries
    data_dictionary['cycles_counts'][key][machine].append(data['core_cycle_counts'][key][p_exec])
AttributeError: 'NoneType' object has no attribute 'append'
```

**Comment by snim2** [comment]
The plot titles without `-b` look like this:

``` python
                    for p_exec in xrange(num_p_execs):
                        if changepoints:
                            classification = ' (%s)' % \
                                data['classifications'][key][p_exec]
                        else:
                            classification = ''
                        title = '%s, %s, %s\nProcess execution #%d%s' % \
                                (benchmark_name,
                                 key.split(':')[1],
                                 machine_name,
                                 p_exec + 1,
                                 classification)
                        plot_titles[key][machine].append(title)
```

And the plot titles with `-b` should also include classifications, if the user has asked for them.

**Comment by vext01** [comment]
I seem to have added the executions twice accidentally...

**Comment by snim2** [comment]
This is working for me now...

**Comment by vext01** [comment]
I _think_ i managed to fix this, but you need to check it carefully and run some data over it with different flags.

This script is out of control. It has become very unmaintainable. I think post-submission we could try to tidy it up. We could really benefit from some data structures, like `Page` and `Plot`. Currently everything is lists and dicts, and it's oh so confusing.

**Comment by snim2** [comment]
I agree. Testing now...

**Comment by snim2** [comment]
OK, this seems to be working for the calls I've tried. The one thing that doesn't work is the problem you and I have both come across recently -- if the user wants _all_ keys from all files, we assume that all files have the same keys. If they don't the script crashes.

One way around this is to slightly refactor the current code. Currently we create a structure with all the keys for `requested_data`, i.e. keys passed in with `-b`. Then we have a big switch statement that says "if the user requested some keys, just marshal data for those, if not collect data from all keys", and the second branch fails. Instead, I think we should say "get the requested keys, which are _either_ the keys passed in with `-b` _or_ all keys from all JSON files. Then we can remove the `if` statement and just collect every key in `requested_data`. 

Does that make sense?

**Comment by vext01** [comment]
I'm not 100% sure what you mean, but if it cleans up this code, sure! Separate PR though?

File: bin/plot_krun_results [hunk 1]
```diff
@@ -124,6 +124,19 @@ EXPORT_SIZE_INCHES = 12, 10
 DPI = 300
 
 
+def add_instr_data(data_dictionary, key, machine):
```
Comment (snim2) [inline bin/plot_krun_results:127]: Maybe `_add_instr_data`

File: bin/plot_krun_results [hunk 2]
```diff
@@ -199,6 +213,8 @@ def main(is_interactive, data_dcts, plot_titles, window_size, outfile,
         all_changepoint_means.append(page_changepoint_means)
     else:  # Create multiple pages.
         for key in sorted(data_dcts['data']):
+            if key.split(":")[0] == "empty":
```
Comment (snim2) [inline bin/plot_krun_results]: Is this debug code?

File: bin/plot_krun_results [hunk 3]
```diff
@@ -816,13 +823,28 @@ def get_data_dictionaries(json_files, benchmarks=[], wallclock_only=False,
                 if key not in data['wallclock_times']:
```
Comment (snim2) [inline bin/plot_krun_results:822]: Around here is the `requested_data` dictionary, btw.

File: bin/plot_krun_results [hunk 4]
```diff
@@ -766,16 +782,7 @@ def get_data_dictionaries(json_files, benchmarks=[], wallclock_only=False,
                         data_dictionary['aperf_counts'][key][machine] = data['aperf_counts'][key]
```
Comment (snim2) [inline bin/plot_krun_results:781]: Just above here, line 759 is the switch `if benchmarks == []:` and that's the if/else I suggest we remove.

**Comment by snim2** [comment]
OK, other than those minor comments this looks good to me. I'll leave you to think about whether removing the if/else should be in this PR or not.

File: bin/plot_krun_results (hunk 1 above)
Comment (vext01) [reply bin/plot_krun_results:127]: It's not like this is designed to imported elsewhere. Or did I miss the point?

File: bin/plot_krun_results (hunk 2 above)
Comment (vext01) [reply bin/plot_krun_results]: Kind of. It's the hack which prevents the plots crashing for the empty benchmarks.

**Comment by vext01** [comment]
I think we should push on, so I can get graphs into the paper.

**Comment by snim2** [comment]
OK

File: bin/plot_krun_results (hunk 2 above)
Comment (vext01) [reply bin/plot_krun_results]: Shall I add a comment at least?

Comment (snim2) [reply bin/plot_krun_results]: They don't crash for me, are you sure they still crash for you?

File: bin/plot_krun_results (hunk 1 above)
Comment (snim2) [reply bin/plot_krun_results:127]: OK, if you prefer

File: bin/plot_krun_results (hunk 2 above)
Comment (vext01) [reply bin/plot_krun_results]:
```Traceback (most recent call last):
  File "/usr/lib/python2.7/pdb.py", line 1314, in main
    pdb._runscript(mainpyfile)
  File "/usr/lib/python2.7/pdb.py", line 1233, in _runscript
    self.run(statement)
  File "/usr/lib/python2.7/bdb.py", line 400, in run
    exec cmd in globals, locals
  File "<string>", line 1, in <module>
  File "/home/vext01/research/warmup_experiment/bin/plot_krun_results", line 4, in <module>
    """
  File "/home/vext01/research/warmup_experiment/bin/plot_krun_results", line 257, in main
    inset_xlimit)
  File "/home/vext01/research/warmup_experiment/bin/plot_krun_results", line 609, in draw_page
    changepoint_mean_exec, median, tukey)
  File "/home/vext01/research/warmup_experiment/bin/plot_krun_results", line 428, in draw_process_exec
    format_yticks_scientific(axis1)
  File "/home/vext01/research/warmup_experiment/warmup/plotting.py", line 22, in format_yticks_scientific
    pyplot.draw()
  File "/usr/lib/python2.7/dist-packages/matplotlib/pyplot.py", line 571, in draw
    get_current_fig_manager().canvas.draw()
  File "/usr/lib/python2.7/dist-packages/matplotlib/backends/backend_agg.py", line 469, in draw
    self.figure.draw(self.renderer)
  File "/usr/lib/python2.7/dist-packages/matplotlib/artist.py", line 59, in draw_wrapper
    draw(artist, renderer, _args, *_kwargs)
  File "/usr/lib/python2.7/dist-packages/matplotlib/figure.py", line 1079, in draw
    func(_args)
  File "/usr/lib/python2.7/dist-packages/matplotlib/artist.py", line 59, in draw_wrapper
    draw(artist, renderer, *args, *_kwargs)
  File "/usr/lib/python2.7/dist-packages/matplotlib/axes/_base.py", line 2092, in draw
    a.draw(renderer)
  File "/usr/lib/python2.7/dist-packages/matplotlib/artist.py", line 59, in draw_wrapper
    draw(artist, renderer, _args, *_kwargs)
  File "/usr/lib/python2.7/dist-packages/matplotlib/axis.py", line 1114, in draw
    ticks_to_draw = self._update_ticks(renderer)
  File "/usr/lib/python2.7/dist-packages/matplotlib/axis.py", line 957, in _update_ticks
    tick_tups = [t for t in self.iter_ticks()]
  File "/usr/lib/python2.7/dist-packages/matplotlib/axis.py", line 903, in iter_ticks
    self.major.formatter.set_locs(majorLocs)
  File "/usr/lib/python2.7/dist-packages/matplotlib/ticker.py", line 523, in set_locs
    self._set_format(vmin, vmax)
  File "/usr/lib/python2.7/dist-packages/matplotlib/ticker.py", line 584, in _set_format
    loc_range_oom = int(math.floor(math.log10(loc_range)))
ValueError: math domain error
Uncaught exception. Entering post mortem debugging
Running 'cont' or 'step' will restart the program

```
```

Comment (snim2) [reply bin/plot_krun_results]: I think this is because you are using an old matplotlib on bencher2. Do you have access to a more recent one?

**Comment by vext01** [comment]
Using a new MPL works. Removed hack. OK to squash?

**Comment by snim2** [comment]
Yup!

**Review by snim2 — APPROVED** [review]

**Pull request merged**